frame/glutton/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+				Error::<T>::AlreadyInitialized
+			);
+
+			// Add or remove elements from `TrashData`.
+			(new_count..current_count).for_each(|i| TrashData::<T>::remove(i));
+			(current_count..new_count).for_each(|i| TrashData::<T>::insert(i, [i as u8; 1024]));
+
+			Self::deposit_event(Event::PalletInitialized { reinit: witness_count.is_some() });
+			TrashDataCount::<T>::set(new_count);
+			Ok(())
+		}
+
+		/// Set the `Compute` storage value that determines how much of the
+		/// block's weight `ref_time` to use during `on_idle`.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::set_compute())]
+		pub fn set_compute(origin: OriginFor<T>, compute: Perbill) -> DispatchResult {
+			ensure_root(origin)?;
+			Compute::<T>::set(compute);
+
+			Self::deposit_event(Event::ComputationLimitSet { compute });
+			Ok(())
+		}
+
+		/// Set the `Storage` storage value that determines the PoV size usage
+		/// for each block.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::set_storage())]
+		pub fn set_storage(origin: OriginFor<T>, storage: Perbill) -> DispatchResult {
+			ensure_root(origin)?;
+			Storage::<T>::set(storage);
+
+			Self::deposit_event(Event::StorageLimitSet { storage });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Waste at most the remaining proof size of `meter`.
+		///
+		/// Tries to come as close to the limit as possible.
+		pub(crate) fn waste_at_most_proof_size(meter: &mut WeightMeter) {
+			let Ok(n) = Self::calculate_proof_size_iters(&meter) else {
+				return;
+			};
+
+			meter.defensive_saturating_accrue(T::WeightInfo::waste_proof_size_some(n));
+
+			(0..n).for_each(|i| {
+				TrashData::<T>::get(i);
+			});
+		}
+
+		/// Calculate how many times `waste_proof_size_some` should be called to fill up `meter`.
+		fn calculate_proof_size_iters(meter: &WeightMeter) -> Result<u32, ()> {
+			let base = T::WeightInfo::waste_proof_size_some(0);
+			let slope = T::WeightInfo::waste_proof_size_some(1).saturating_sub(base);
+
+			let remaining = meter.remaining().saturating_sub(base);
+			let iter_by_proof_size =
+				remaining.proof_size().checked_div(slope.proof_size()).ok_or(())?;
+			let iter_by_ref_time = remaining.ref_time().checked_div(slope.ref_time()).ok_or(())?;
+
+			match iter_by_proof_size {
+				0 => Err(()),
+				i if i <= iter_by_ref_time => Ok(i as u32),
+				_ => Err(()),
+			}
+		}
+
+		/// Waste at most the remaining ref time weight of `meter`.
+		///
+		/// Tries to come as close to the limit as possible.
+		pub(crate) fn waste_at_most_ref_time(meter: &mut WeightMeter) {
+			let Ok(n) = Self::calculate_ref_time_iters(&meter) else {
+				return;
+			};
+			meter.defensive_saturating_accrue(T::WeightInfo::waste_ref_time_iter(n));
+
+			let clobber = Self::waste_ref_time_iter(vec![0u8; 64], n);
+
+			// By casting it into a vec we can hopefully prevent the compiler from optimizing it
```

Comment:
Usually you'd use something like `std::ptr::read_volatile` to prevent the compiler from optimizing a variable out. (These reads are guaranteed to not be elided or reordered by the compiler.)

Reply:
Interesting. Didn't know about this feature in rust. IIUC this is an unsafe function but since we are not using `std::ptr::write_volatile` this is kind of safe. But I still think we are better off here using our current approach. @ggwpez Thoughts?

Reply:
Yea I would rather not put `unsafe` into the runtime code - even if that is the more correct way.   
I think it should be fine in this case, or?

Reply:
Yes, the code should be also okay the way it is. `unsafe` isn't that bad and we also use `unsafe` in the runtime (but just hidden :P).

Reply:
The way to go would probably be to just make an utility function somewhere that would wrap the unsafe call and make 100% sure it's always safe. There's at least one more place that I know of where this would be useful, if not more. :P

The thing with `unsafe` isn't that it should never be used, it's just that its use should be limited (it shouldn't be sprinkled all over the code), neatly encapsulated in a safe interface, and very carefully written and reviewed (ideally by experts).

But yeah, if it works then the way it is now is fine for now. It was more of a drive-by comment rather than a "hey, change this".